Add support for get_work_dim() and global offset

kpet opened the pull request:
Use push constants to pass the values from the runtime/application.
The required push constants are declared in the descriptor map.

- Introduce generic support for push constants
- Move work-item built-in function tests to a separate folder
- Introduce validation code for explicit structure layouts
- Add -scalar-block-layout option to be able to change the behaviour
  of layout checks based on the presence of the feature
- Fix a few typos in related code.

Contributes to #341, #342.

Signed-off-by: Kévin Petit <kpet@free.fr>

alan-baker commented on lib/DefineOpenCLWorkItemBuiltinsPass.cpp:
```diff
@@ -212,16 +263,36 @@ bool DefineOpenCLWorkItemBuiltinsPass::defineGlobalSizeBuiltin(Module &M) {
+  // If global offset support is disabled and the builtin not used in the
+  // module, don't create it. Otherwise, always define it as it's used in global
+  // ID calculations.
+  if (!clspv::Option::GlobalOffset() && (nullptr == F)) {
```
get_global_id only uses get_global_offset if `clspv::Option::GlobalOffset()` returns true. It seems like this condition shouldn't define the function unless the option is set.

kpet replied:
This is actually what the code does but reversing the logic might make this clearer.

kpet replied:
I've reworked the code a bit to avoid re-declaring `get_global_offset` when it's already used and reworded the comments a bit. Hopefully this makes things clearer.

alan-baker commented on lib/Option.cpp:
```diff
@@ -168,6 +168,18 @@ llvm::cl::opt<clspv::Option::SourceLanguage> cl_std(
+    llvm::cl::desc("Enable support for get_work_dim() built-in function"));
+
+static llvm::cl::opt<bool>
+    global_offset("global-offset", llvm::cl::init(false),
```
I assume this is intended to default on at some point in the future (e.g. when support is landed in clvk)?

kpet replied:
Maybe but that would mean all the applications that use clspv have to support and pass push constants.

alan-baker replied:
Only if those builtins are used. I don't think that's a problem since push constants are a core feature.

kpet replied:
Currently the descriptor map entries are always present but we could only emit them when the built-in is used for `get_work_dim`. For `get_global_offset` however, the entry would always be present if `get_global_id` is used. I'm not worried about optionality, just forcing applications to support something they may not need. Having said that, using push constants for POD arguments might be a good justification for requiring applications to pass push constants.

alan-baker commented on test/WorkItemBuiltins/multiple-push-constant.cl:
```diff
@@ -0,0 +1,34 @@
+// RUN: FileCheck --check-prefix=DMAP %s < %t.dmap
+// RUN: spirv-val --target-env vulkan1.0 %t.spv
+
+// DMAP: pushconstant,name,dimensions,offset,0,size,4
```
You should flip these element to generate:
```
// DMAP: pushconstant,name,global_offset,offset,0,size16
// DMAP: pushconstant,name,dimensions,offset,12,size,4
```

kpet replied:
Doh! I should have spotted that, thanks.

alan-baker commented on lib/PushConstant.h:
```diff
@@ -0,0 +1,31 @@
+
+namespace clspv {
+
+const char *GetPushConstantName(PushConstant);
```
Function comments please.

kpet replied:
Done.

alan-baker commented on test/WorkItemBuiltins/get_global_id-with-global-offset.cl:
```diff
@@ -0,0 +1,31 @@
+// RUN: clspv -global-offset %s -o %t.spv
```
Can you please add a version of this test for:
```
out[0] = get_global_id(3);
```

kpet replied:
Done.

alan-baker commented on test/WorkItemBuiltins/get_global_offset-push-constant.cl:
```diff
@@ -0,0 +1,27 @@
+// RUN: clspv -global-offset -descriptormap=%t.dmap %s -o %t.spv
```
Same here, a version for an out of range access.

kpet replied:
That is covered by the `get_global_id(3)` test above IMHO since `get_global_id` calls `get_global_offset`. The only thing that makes it into the SPIR-V is a store of a 0 constant.

alan-baker commented on test/WorkItemBuiltins/multiple-push-constant.cl:
```diff
@@ -0,0 +1,34 @@
+// RUN: clspv -work-dim -global-offset -descriptormap=%t.dmap %s -o %t.spv
```
Another variant of this test please, call get_global_offset with a larger index (or non-constant index) to generate the select.

kpet replied:
I've ended up doing another variant of the simple `get_global_offset` test with a non-constant index. I didn't feel having get_work_dim in the same test added anything.

alan-baker commented on lib/SPIRVProducerPass.cpp:
```diff
@@ -2850,6 +2859,230 @@ void SPIRVProducerPass::GenerateResourceVars(Module &) {
+  llvm_unreachable("Unsupported type");
+}
+
+unsigned roundUpToMultipleOf(unsigned num, unsigned multiple) {
```
This is effectively MathExtras.h `alignTo` (with Skew = 0). I'd prefer reusing that to adding another definition.

kpet replied:
Thanks! I was sure there was already a function for this somewhere.

alan-baker commented on lib/DeclarePushConstantsPass.cpp:
```diff
@@ -0,0 +1,100 @@
+
+  auto &C = M.getContext();
+
+  if (clspv::Option::WorkDim()) {
```
Swap these as mentioned in the multiple push constants test.

kpet replied:
Done.

kpet commented:
Thanks for the review!

>We'll need an argument to pass the maxPushConstantsSize limit to the compiler.

That would be good but we can also assume the minimum size required by the Vulkan spec.

I think I've addressed the review comments and fixed the Windows build error. I can't make sense of the clvk bot failure.

PTAl.

alan-baker commented:
Don't worry about the clvk bot. A recent swiftshader commit seems to have broken it for that use.

I think for the argument I would just default to the minimum maxPushConstantsSize and let users (including clvk) override it as they wish. It's for the future anyways.
